hera_cal/lstbin_simple.py

```diff
@@ -0,0 +1,1740 @@
+        freqs = freqs[mask]
+        freq_chans = freq_chans[mask]
+
+    if pols is None:
+        pols = metas[0].pols
+    else:
+        if not all(isinstance(p, str) for p in pols):
+            pols = uvutils.polnum2str(pols, x_orientation=metas[0].x_orientation)
+
+    if antpos is None and rephase:
+        warnings.warn(
+            "Getting antpos from the first file only. This is almost always correct, "
+            "but will be wrong if different files have different antenna_position arrays."
+        )
+        antpos = dict(zip(metas[0].antenna_numbers, metas[0].antpos_enu))
+
+    if time_idx is None:
+        adjust_lst_bin_edges(lst_bin_edges)        
+        lst_bin_edges %= 2*np.pi
+        op = np.logical_and if lst_bin_edges[0] < lst_bin_edges[-1] else np.logical_or
+        time_idx = []
+        for meta in metas:
+            _lsts = meta.get_transactional('lsts')
+            time_idx.append(op(_lsts >= lst_bin_edges[0], _lsts < lst_bin_edges[-1]))
+
+    if time_arrays is None:
+        time_arrays = [meta.get_transactional('times')[idx] for meta, idx in zip(metas, time_idx)]
+
+    if lsts is None:
+        lsts = np.concatenate(
+            [meta.get_transactional('lsts')[idx] for meta, idx in zip(metas, time_idx)]
+        )            
+
+    # Now we can set up our master arrays of data. 
+    data, flags, nsamples = _allocate_dnf(
+        (len(lsts), len(antpairs), len(freqs), len(pols)),
+        d=np.nan + np.nan*1j,
+        f=True
+    )
+
+    if cal_files is None:
+        cal_files = [None] * len(metas)
+
+    if redundantly_averaged and reds is None:
+        raise ValueError("reds must be provided if redundantly_averaged is True")
+    if redundantly_averaged and any(c is not None for c in cal_files):
+        raise ValueError("Cannot apply calibration if redundantly_averaged is True")
+    
+    # This loop actually reads the associated data in this LST bin.
+    ntimes_so_far = 0
+    for meta, calfl, tind, tarr in zip(metas, cal_files, time_idx, time_arrays):
+        logger.info(f"Reading {meta.path}")
+        slc = slice(ntimes_so_far,ntimes_so_far+len(tarr))
+        ntimes_so_far += len(tarr)
+
+        #hd = io.HERAData(str(fl.path), filetype='uvh5')
+        data_antpairs = meta.get_transactional('antpairs')
+
+        if redundantly_averaged:
+            bls_to_load = [bl for bl in data_antpairs if reds.get_ubl_key(bl) in antpairs]
+        else:
+            bls_to_load = [bl for bl in antpairs if bl in data_antpairs or bl[::-1] in data_antpairs]
+
+        if not bls_to_load or not np.any(tind):
+            # If none of the requested baselines are in this file, then just 
+            # set stuff as nan and go to next file. 
+            logger.warning(f"None of the baseline-times are in {meta.path}. Skipping.")
+            data[slc] = np.nan
+            flags[slc] = True
+            nsamples[slc] = 0
+            continue
+
+        # TODO: use Fast readers here instead.
```

Comment:
So the catch is that the `HERADataFastReader` has no ability to down-select on times, which I need. I had tried to add this functionality in this branch, but it was not working quite right somehow, so I gave up. There will be a more robust solution coming soon (given pending PRs on pyuvdata) so I decided to wait on those. This will definitely be something I'll come back to.